File: fitting/nonlinear/c/nonlinfit.c
//Normal distribution fitting code using Levenberg-Marquard method
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define PI  3.14159265359

//Fitting function Normal distribution
//func(x, mu, sig) = Exp[-(x-mu)^2/(2*sig^2)]//(sig*Sqrt[2 * Pi])
float func(float x, float *a, int n) {
  float func;

  int i;
  float mu, sig;
  
  //a[0] = mu
  //a[1] = sig
  mu = a[0];
  sig = a[1];

  func = exp(-(x-mu)*(x-mu)/(2*sig*sig))/(sig*sqrt(2*PI));
  return func;
}

//Function to calculate Chi2
float chi2(float *x, float *y, float *sig, int n, float *a, int nprm) {
  float chi2;
  float dy;
  int i;

  chi2 = 0.0;
  for (i = 0; i < n; i++) {
    dy = y[i] - func(x[i], a, nprm);
    chi2 += dy * dy / (sig[i]*sig[i]);
  }

  return chi2;

}

void dyda(float x, float *a, int nsize, float *res) {
  int i, j;
  float tmp, mu, sig;

  for (i = 0; i < nsize; i++)
    res[i] = 0.0;

  mu  = a[0];
  sig = a[1];
  
  res[0] = (mu - x) * func(x, a, nsize);
  res[1] = -( 1 + (mu-x)*(mu-x)/(sig*sig)) * func(x, a, nsize) / sig;  
}

int main() {
  //Data array
  float *x, *y, *sig;
  //work variables
  int n, i, j, k, nprm;
  
  float da[2], a[2], anew[2], dchi2[2], beta[2], hess[2][2];
  float lambda, chi2cur, chi2new;
  int c1, c2, pivot[2], ok;
  int cnt, maxit;

  fscanf(stdin,"%d",&n);
  
  if (n < 2) {
    fprintf(stderr,"Bad number of records %d\n",n);
    return 1;
  } 

  a[0] = 195;
  a[1] = 45.0;
  nprm = 2;

  x = (float *) malloc(sizeof(float) * n);
  y = (float *) malloc(sizeof(float) * n);
  sig = (float *) malloc(sizeof(float) * n);

  
  for (i = 0; i < n; i++) {
    fscanf(stdin,"%f",&x[i]);
    fscanf(stdin,"%f",&y[i]);
    sig[i] = 1.0;
  }

  //Begin parametrization search
  //
  //calculate da, solve Hess x da = Beta and store result into Beta
  c1 = nprm;
  c2 = 1;
  sgesv_(&c1, &c2, hess, &c1, pivot, beta, &c1, &ok);
  //
  //
  //End parametrization search

  free(x);
  free(y);
  free(sig);

  return 0;

}

File: fitting/nonlinear/fortran/nonlinfit.f90
program nonlinfit
implicit none

double precision, parameter :: PI = 3.14159265359d0

double precision, dimension(:), allocatable :: x, y, sig
double precision, dimension(2) :: da, a, anew, dchi2, beta
double precision, dimension(2,2) :: hess
double precision :: lambda, chi2cur, chi2new

integer n, i, j, k, nprm
integer cnt, maxit, c1, c2, ok
integer :: pivot(2)

interface
function chi2(x, y, sig, n, a, nprm)
double precision :: chi2
double precision, dimension(:), intent(in) :: x, y, sig
integer, intent(in) :: n, nprm
double precision, dimension(:), intent(in) :: a
end function chi2
end interface

read(*,*)n

if ( n < 2) then
        write(0,*) 'Bad number of records = ',n
        stop
end if

a(1) = 195.e0
a(2) = 40.e0

nprm = 2

allocate(x(n))
allocate(y(n))
allocate(sig(n))

do i = 1, n
        read(*,*)x(i),y(i)
        sig(i) = 1.e0
end do

chi2cur = chi2(x, y, sig, n, a, nprm)

lambda = 1.d-5
cnt = 0
maxit = 5

do while ( cnt < maxit)
        cnt = cnt + 1
        !calc the hess matrix
        do j = 1, nprm
                
                do i = 1, nprm
                        hess(i,j) = 0.d0
                        
                end do
        end do

end do


deallocate(x)
deallocate(y)
deallocate(sig)


end program nonlinfit

function chi2(x, y, sig, n, a, nprm)
double precision :: chi2
double precision, dimension(:), intent(in) :: x, y, sig
integer, intent(in) :: n, nprm
double precision, dimension(:), intent(in) :: a

double precision, dimension(:), allocatable :: dy, res

double precision :: mu, sigg
!interface
!elemental function func(x, a, n) result(res)
!double precision :: res

!double precision, intent(in) :: x
!double precision, dimension(:), intent(in) ::  a
!integer, intent(in) :: n
!end function
!end interface

allocate(dy(n))
allocate(res(n))

mu = a(1)
sigg = a(2)

res = exp( -(x - mu)**2/(2*sigg**2) ) / (sigg * sqrt(2*PI))

dy(:) = y(:) - res

chi2 = sum(dy(:) * dy(:) / sig(:) * sig(:) )

deallocate(dy)
deallocate(res)

end function chi2

!elemental function func(x, a, n) result(res)
!double precision :: res

!double precision, intent(in) :: x
!double precision, dimension(:), intent(in) ::  a
!integer, intent(in) :: n



!double precision :: mu, sig
!double precision, parameter :: PI = 3.14159265359d0

!mu = a(1)
!sig = a(2)

!res = exp( -(x - mu)**2/(2*sig**2) ) / (sig * sqrt(2*PI))

!end function func

subroutine dyda(x, a, nsize, res)
double precision :: x
double precision, dimension(:), intent(in) :: a
integer, intent(in) :: nsize
double precision, dimension(nsize), intent(out) :: res

double precision :: mu, sig

res(:) = 0.d0
mu  = a(1)
sigg = a(2)

res = exp( -(x - mu)**2/(2*sigg**2) ) / (sigg * sqrt(2*PI))

res(1) = sum((x - mu) * res / sigg ** 2)
res(2) = - sum(( 1 + (mu - x) ** 2 / sig ** 2) * res / sigg)

end subroutine dyda
